contracts/contracts/rem.attr/src/rem.attr.cpp

```diff
@@ -0,0 +1,115 @@
+#include <rem.attr/rem.attr.hpp>
+
+
+namespace eosio {
+
+   void attribute::confirm( const name& owner, const name& attribute_name )
+   {
+      require_auth(owner);
+
+      attributes_table attributes( _self, owner.value );
+      const auto& attr = attributes.get( attribute_name.value, "account does not have this attribute" );
+      attributes.modify( attr, same_payer, [&]( auto& attr ) {
+         attr.data = attr.pending;
+         attr.pending.clear();
+      });
+   }
+
+   void attribute::create( const name& attribute_name, int32_t type, int32_t ptype )
+   {
+      require_auth( _self );
+      check( type >= 0 && type < static_cast<int32_t>( data_type::MaxVal ), "attribute type is out of range" );
+      check( ptype >= 0 && ptype < static_cast<int32_t>( privacy_type::MaxVal ), "attribute privacy type is out of range" );
+
+      attribute_info_table attributes_info( _self, attribute_name.value );
+      check( attributes_info.find( attribute_name.value ) == attributes_info.end(), "attribute with this name already exists" );
+
+      attributes_info.emplace( _self, [&]( auto& attr ) {
+         attr.attribute_name = attribute_name;
+         attr.type           = type;
+         attr.ptype          = ptype;
+      });
+   }
+
+   void attribute::setattr( const name& issuer, const name& receiver, const name& attribute_name, const std::vector<char>& value )
+   {
+      require_auth( issuer );
+      check( !value.empty(), "value is empty" );
+
+      attribute_info_table attributes_info( _self, attribute_name.value );
+      const auto& attrinfo = attributes_info.get( attribute_name.value, "attribute does not exist" );
+      check_create_permission(issuer, receiver, attrinfo.ptype);
+
+      require_recipient( receiver );
+
+      attributes_table attributes( _self, receiver.value );
+      const auto attr_it = attributes.find( attribute_name.value );
+      if ( attr_it == attributes.end() ) {
+         attributes.emplace( issuer, [&]( auto& attr ) {
+            attr.attribute_name = attribute_name;
+            attr.issuer         = issuer;
+            if (need_confirm(attrinfo.ptype)) {
+               attr.pending = value;
+            }
+            else {
+               attr.data = value;
+            }
+         });
+      } else {
+         check_delete_permission(attr_it->issuer, issuer, receiver, attrinfo.ptype);
```

Comment:
@me-vadim why do we need to check delete permission?

Reply:
@mark-tsyrulnyk because when attribute is already set, 'unset' (or in this case reset) can be executed by account that assigned this attribute or receiver account (if attribute needs confirmation)

Reply:
Reworked logic, removed check_delete_permission